### PR TITLE
Fix template return missing a value

### LINKED ISSUE #60: The null object does not have a method 'visitExpression'
**Overview:** When running `pub build`, compilation fails while compiling templates. Looking at the stack trace, it appears that the error occurs when visiting a `return` statement. For some reason, the AST's representation of the statement has a null `value`. Here is the stack trace:

```
Working dir: /Users/tobe/WebstormProjects/proxy-dispenser
/usr/local/opt/dart/libexec/bin/pub build --mode=release --output=build
Loading source assets...
Loading dart_to_js_script_rewriter, angular2/transform/codegen, observe and smoke/src/default_transformer transformers...
Loading angular2 transformers...
Building angel...
Build error:
Transform TemplateCompiler on angel|web/main.ng_meta.json threw error: The null object does not have a method 'visitExpression'.

NoSuchMethodError: method not found: 'visitExpression'
Receiver: null
Arguments: [Instance of '_DartEmitterVisitor', Instance of 'EmitterVisitorContext']
dart:core-patch/object_patch.dart 44                                              Object._noSuchMethod
dart:core-patch/object_patch.dart 47                                              Object.noSuchMethod
package:angular2/src/compiler/output/abstract_emitter.dart 131:16                 AbstractEmitterVisitor.visitReturnStmt
package:angular2/src/compiler/output/output_ast.dart 570:20                       ReturnStatement.visitStatement
package:angular2/src/compiler/output/abstract_emitter.dart 469:19                 AbstractEmitterVisitor.visitAllStatements.<fn>
dart:core-patch/growable_array.dart 258                                           List.forEach
package:angular2/src/compiler/output/abstract_emitter.dart 468:16                 AbstractEmitterVisitor.visitAllStatements
package:angular2/src/compiler/output/dart_emitter.dart 187:10                     _DartEmitterVisitor._visitClassMethod
package:angular2/src/compiler/output/dart_emitter.dart 122:43                     _DartEmitterVisitor.visitDeclareClassStmt.<fn>
dart:core-patch/growable_array.dart 258                                           List.forEach
package:angular2/src/compiler/output/dart_emitter.dart 122:18                     _DartEmitterVisitor.visitDeclareClassStmt
package:angular2/src/compiler/output/output_ast.dart 631:20                       ClassStmt.visitStatement
package:angular2/src/compiler/output/abstract_emitter.dart 469:19                 AbstractEmitterVisitor.visitAllStatements.<fn>
dart:core-patch/growable_array.dart 258                                           List.forEach
package:angular2/src/compiler/output/abstract_emitter.dart 468:16                 AbstractEmitterVisitor.visitAllStatements
package:angular2/src/compiler/output/dart_emitter.dart 53:15                      DartEmitter.emitStatements
package:angular2/src/compiler/offline_compiler.dart 143:14                        OfflineCompiler._codegenSourceModule
package:angular2/src/compiler/offline_compiler.dart 106:12                        OfflineCompiler.compile
package:angular2/src/transform/template_compiler/generator.dart 57:29             processTemplates.<async>.<fn>
package:angular2/src/transform/common/logging.dart 37:27                          logElapsedSync
package:angular2/src/transform/template_compiler/generator.dart 56:29             processTemplates.<async>
dart:async-patch/async_patch.dart 27                                              _asyncThenWrapperHelper.<fn>
dart:async/zone.dart 1158                                                         _rootRunUnary
dart:async/zone.dart 1037                                                         _CustomZone.runUnary
dart:async/future_impl.dart 131                                                   _FutureListener.handleValue
dart:async/future_impl.dart 637                                                   _Future._propagateToListeners.handleValueCallback
dart:async/future_impl.dart 667                                                   _Future._propagateToListeners
dart:async/future_impl.dart 467                                                   _Future._complete
dart:async/future_impl.dart 52                                                    _SyncCompleter.complete
package:angular2/src/transform/common/logging.dart 27:3                           logElapsedAsync.<async>
dart:async-patch/async_patch.dart 27                                              _asyncThenWrapperHelper.<fn>
dart:async/zone.dart 1158                                                         _rootRunUnary
dart:async/zone.dart 1037                                                         _CustomZone.runUnary
dart:async/future_impl.dart 131                                                   _FutureListener.handleValue
dart:async/future_impl.dart 637                                                   _Future._propagateToListeners.handleValueCallback
dart:async/future_impl.dart 667                                                   _Future._propagateToListeners
dart:async/future_impl.dart 467                                                   _Future._complete
dart:async/future_impl.dart 52                                                    _SyncCompleter.complete
package:angular2/src/transform/template_compiler/compile_data_creator.dart 121:5  _CompileDataCreator.createCompileData.<async>
dart:async-patch/async_patch.dart 27                                              _asyncThenWrapperHelper.<fn>
dart:async/zone.dart 1158                                                         _rootRunUnary
dart:async/zone.dart 1037                                                         _CustomZone.runUnary
dart:async/future_impl.dart 131                                                   _FutureListener.handleValue
dart:async/future_impl.dart 637                                                   _Future._propagateToListeners.handleValueCallback
dart:async/future_impl.dart 667                                                   _Future._propagateToListeners
dart:async/future_impl.dart 467                                                   _Future._complete
dart:async/future_impl.dart 52                                                    _SyncCompleter.complete
package:angular2/src/transform/template_compiler/compile_data_creator.dart 268:5  _CompileDataCreator._extractNgMeta.<async>
dart:async-patch/async_patch.dart 27                                              _asyncThenWrapperHelper.<fn>
dart:async/zone.dart 1158                                                         _rootRunUnary
dart:async/zone.dart 1037                                                         _CustomZone.runUnary
dart:async/future_impl.dart 131                                                   _FutureListener.handleValue
dart:async/future_impl.dart 637                                                   _Future._propagateToListeners.handleValueCallback
dart:async/future_impl.dart 667                                                   _Future._propagateToListeners
dart:async/future_impl.dart 485                                                   _Future._completeError
dart:async/future_impl.dart 537                                                   _Future._asyncCompleteError.<fn>
dart:async/zone.dart 1150                                                         _rootRun
dart:async/zone.dart 1026                                                         _CustomZone.run
dart:async/zone.dart 953                                                          _CustomZone.bindCallback.<fn>
dart:async/schedule_microtask.dart 41                                             _microtaskLoop
dart:async/schedule_microtask.dart 50                                             _startMicrotaskLoop
dart:isolate-patch/isolate_patch.dart 96                                          _runPendingImmediateCallback
dart:isolate-patch/isolate_patch.dart 149                                         _RawReceivePortImpl._handleMessage

dart:core                                                                         Object.noSuchMethod
package:angular2/src/compiler/output/abstract_emitter.dart 131:16                 AbstractEmitterVisitor.visitReturnStmt
package:angular2/src/compiler/output/output_ast.dart 570:20                       ReturnStatement.visitStatement
package:angular2/src/compiler/output/abstract_emitter.dart 469:19                 AbstractEmitterVisitor.visitAllStatements.<fn>
dart:core                                                                         List.forEach
package:angular2/src/compiler/output/abstract_emitter.dart 468:16                 AbstractEmitterVisitor.visitAllStatements
package:angular2/src/compiler/output/dart_emitter.dart 187:10                     _DartEmitterVisitor._visitClassMethod
package:angular2/src/compiler/output/dart_emitter.dart 122:43                     _DartEmitterVisitor.visitDeclareClassStmt.<fn>
dart:core                                                                         List.forEach
package:angular2/src/compiler/output/dart_emitter.dart 122:18                     _DartEmitterVisitor.visitDeclareClassStmt
package:angular2/src/compiler/output/output_ast.dart 631:20                       ClassStmt.visitStatement
package:angular2/src/compiler/output/abstract_emitter.dart 469:19                 AbstractEmitterVisitor.visitAllStatements.<fn>
dart:core                                                                         List.forEach
package:angular2/src/compiler/output/abstract_emitter.dart 468:16                 AbstractEmitterVisitor.visitAllStatements
package:angular2/src/compiler/output/dart_emitter.dart 53:15                      DartEmitter.emitStatements
package:angular2/src/compiler/offline_compiler.dart 143:14                        OfflineCompiler._codegenSourceModule
package:angular2/src/compiler/offline_compiler.dart 106:12                        OfflineCompiler.compile
package:angular2/src/transform/template_compiler/generator.dart 57:29             processTemplates.<async>.<fn>
package:angular2/src/transform/common/logging.dart 37:27                          logElapsedSync
package:angular2/src/transform/template_compiler/generator.dart 56:29             processTemplates.<async>
dart:async                                                                        _SyncCompleter.complete
package:angular2/src/transform/common/logging.dart 27:3                           logElapsedAsync.<async>
dart:async                                                                        _SyncCompleter.complete
package:angular2/src/transform/template_compiler/compile_data_creator.dart 121:5  _CompileDataCreator.createCompileData.<async>
dart:async                                                                        _SyncCompleter.complete
package:angular2/src/transform/template_compiler/compile_data_creator.dart 268:5  _CompileDataCreator._extractNgMeta.<async>
[Info from Dart2JS]:
Compiling angel|web/main.dart...
[Error from Dart2JS]:
web/main.dart:3:1:
Can't read 'file:///Users/tobe/WebstormProjects/proxy-dispenser/web/main.template.dart' (Could not find asset angel|web/main.template.dart.).
import 'main.template.dart' as ngStaticInit;import "dart:convert" show JSON;
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[Warning from Dart2JS]:
web/main.dart:26:135:
'ngStaticInit' has no member named 'initReflector'.
  bootstrapStatic(ProxyAppComponent, [ROUTER_PROVIDERS, provide(LocationStrategy, useClass: HashLocationStrategy), UserService], () { ngStaticInit.initReflector(); });
                                                                                                                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Build failed.
[Info from Dart2JS]:
Took 0:00:06.025756 to compile angel|web/main.dart.
Process finished with exit code 65
```

And the code that ostensibly causes it:

``` dart
dynamic visitReturnStmt(o.ReturnStatement stmt, dynamic context) {
  EmitterVisitorContext ctx = context;
  ctx.print('''return ''');

  // stmt.value is null!!!
  stmt.value.visitExpression(this, ctx);
  ctx.println(";");
  return null;
}
```

**Angular Version:** Occurs in `2.0.0-beta.18` and `2.0.0-beta.20`. I haven't tried this in other versions.

**Motivation:** I am trying to launch an app, and I cannot launch it if the source cannot be compiled to JS. 😢 

**Browsers:** N/A

**Reproduce the Error:** I am not sure. All I did was run `pub build`.

**Related Issues:** None

**Suggest a Fix:** Not yet sure what's causing this, so I have no suggestions.

Thanks in advance. Also, if there is any way I can contribute to Dart or any of these projects, I would very much enjoy it. I have made some cool packages with Dart, and am working on a [Web server framework](https://github.com/angel-dart/angel). In fact, this project uses it.

Best wishes!

### PR DESCRIPTION
**Pull request by thosakwe**
I want to resolve #60 by enabling support for return statements without a returned expression. 😄

### CODE REVIEW

**Comment by thosakwe** [comment]
I signed it!

File: lib/src/compiler/output/abstract_emitter.dart [hunk 1]
```diff
@@ -125,7 +125,8 @@ abstract class AbstractEmitterVisitor
   dynamic visitReturnStmt(o.ReturnStatement stmt, dynamic context) {
     EmitterVisitorContext ctx = context;
     ctx.print('''return ''');
-    stmt.value.visitExpression(this, ctx);
+
+    if (stmt.value != null) stmt.value.visitExpression(this, ctx);
```
Comment (kevmoo) [inline lib/src/compiler/output/abstract_emitter.dart]: Better –
```dart
stmt.value?.visitExpression(this, ctx);
```

**Comment by kevmoo** [comment]
@thosakwe can you write a test that reproduces this issue?

**Comment by kevmoo** [comment]
@matanlurey @ferhatb Should we have tests here?

**Comment by thosakwe** [comment]
Hi, I was looking at this again, and source of the problem is not what I guessed. Indeed, there is a `return` statement without an expression, but it doesn't seem to be affected by whether I return void in my app code. It seems like there is something within the project that is causing this. I'll keep looking so I can help out.

In the meantime, I will close this PR.

**Pull request closed without merging**